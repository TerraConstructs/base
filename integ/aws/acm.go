package aws

import (
	"fmt"
	"time"

	"github.com/aws/aws-sdk-go/service/acm"
	"github.com/stretchr/testify/require"

	"github.com/gruntwork-io/terratest/modules/aws"
	"github.com/gruntwork-io/terratest/modules/logger"
	"github.com/gruntwork-io/terratest/modules/retry"
	"github.com/gruntwork-io/terratest/modules/testing"
)

// ref: https://github.com/gruntwork-io/terratest/blob/v0.47.1/modules/aws/acm.go

// Get Certificate Status
func GetAcmCertificateStatus(t testing.TestingT, awsRegion string, certArn string) string {
	status, err := GetAcmCertificateStatusE(t, awsRegion, certArn)
	if err != nil {
		t.Fatal(err)
	}
	return status
}

// GetAcmCertificateStatusE gets the ACM certificate status for the given certificate ARN in the given region.
func GetAcmCertificateStatusE(t testing.TestingT, awsRegion string, certArn string) (string, error) {
	acmClient, err := aws.NewAcmClientE(t, awsRegion)
	if err != nil {
		return "", err
	}

	result, err := acmClient.DescribeCertificate(&acm.DescribeCertificateInput{
		CertificateArn: &certArn,
	})
	if err != nil {
		return "", err
	}

	return *result.Certificate.Status, nil
}

// WaitForCertificateIssued waits for the certificate to be issued
func WaitForCertificateIssued(
	t testing.TestingT,
	certArn string,
	region string,
	maxRetries int,
	sleepBetweenRetries time.Duration,
) {
	err := WaitForCertificateIssuedE(t, certArn, region, maxRetries, sleepBetweenRetries)
	require.NoError(t, err)
}

// WaitForCertificateIssuedE waits for the ACM Certificate to be issued
func WaitForCertificateIssuedE(
	t testing.TestingT,
	certArn string,
	region string,
	maxRetries int,
	sleepBetweenRetries time.Duration,
) error {
	msg, err := retry.DoWithRetryE(
		t,
		fmt.Sprintf("Waiting for Certificate %s to be ISSUED.", certArn),
		maxRetries,
		sleepBetweenRetries,
		func() (string, error) {
			certStatus, err := GetAcmCertificateStatusE(t, region, certArn)
			if err != nil {
				return "", err
			}
			if certStatus != acm.CertificateStatusIssued {
				return "", NewCertificateNotIssuedError(certArn, certStatus)
			}
			return fmt.Sprintf("Certificate %s is now at desired status %s", certArn, acm.CertificateStatusIssued), nil
		},
	)
	logger.Log(t, msg)
	return err
}
