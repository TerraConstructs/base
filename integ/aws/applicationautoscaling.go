package aws

import (
	"context"
	"fmt"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/applicationautoscaling"
	"github.com/aws/aws-sdk-go-v2/service/applicationautoscaling/types"
	"github.com/stretchr/testify/require"

	terratestaws "github.com/gruntwork-io/terratest/modules/aws"
	"github.com/gruntwork-io/terratest/modules/testing"
)

// GetTableTrackingPolicy gets the target tracking policy for a DynamoDB table or errors if not found
func GetTableTrackingPolicy(t testing.TestingT, awsRegion string, resourceId string) *types.ScalingPolicy {
	policy, err := GetTableTrackingPolicyE(t, awsRegion, resourceId)
	require.NoError(t, err)
	return policy
}

// GetTableTrackingPolicy gets the target tracking policy for a DynamoDB table or returns an error if not found
func GetTableTrackingPolicyE(t testing.TestingT, awsRegion string, resourceId string) (*types.ScalingPolicy, error) {
	policies := GetScalingPolicies(t, awsRegion, "dynamodb")

	var targetTrackingPolicy *types.ScalingPolicy
	for _, policy := range policies {
		if *policy.ResourceId == resourceId &&
			policy.ScalableDimension == types.ScalableDimensionDynamoDBTableReadCapacityUnits &&
			policy.PolicyType == types.PolicyTypeTargetTrackingScaling {
			targetTrackingPolicy = &policy
			break
		}
	}
	if targetTrackingPolicy == nil {
		return nil, fmt.Errorf("no target tracking policy found for resource ID: %s", resourceId)
	}
	return targetTrackingPolicy, nil
}

// GetScalableTargets gets the Application Auto Scaling scalable targets for the given service namespace
func GetScalableTargets(t testing.TestingT, region string, serviceNamespace string) []types.ScalableTarget {
	targets, err := GetScalableTargetsE(t, region, serviceNamespace)
	require.NoError(t, err)
	return targets
}

// GetScalableTargetsE gets the Application Auto Scaling scalable targets for the given service namespace
func GetScalableTargetsE(t testing.TestingT, region string, serviceNamespace string) ([]types.ScalableTarget, error) {
	client, err := NewApplicationAutoScalingClientE(t, region)
	if err != nil {
		return nil, err
	}

	result, err := client.DescribeScalableTargets(context.Background(), &applicationautoscaling.DescribeScalableTargetsInput{
		ServiceNamespace: types.ServiceNamespace(serviceNamespace),
	})

	if err != nil {
		return nil, err
	}

	return result.ScalableTargets, nil
}

// GetScalingPolicies gets the Application Auto Scaling scaling policies for the given service namespace
func GetScalingPolicies(t testing.TestingT, region string, serviceNamespace string) []types.ScalingPolicy {
	policies, err := GetScalingPoliciesE(t, region, serviceNamespace)
	require.NoError(t, err)
	return policies
}

// GetScalingPoliciesE gets the Application Auto Scaling scaling policies for the given service namespace
func GetScalingPoliciesE(t testing.TestingT, region string, serviceNamespace string) ([]types.ScalingPolicy, error) {
	client, err := NewApplicationAutoScalingClientE(t, region)
	if err != nil {
		return nil, err
	}

	result, err := client.DescribeScalingPolicies(context.Background(), &applicationautoscaling.DescribeScalingPoliciesInput{
		ServiceNamespace: types.ServiceNamespace(serviceNamespace),
	})

	if err != nil {
		return nil, err
	}

	return result.ScalingPolicies, nil
}

// GetScheduledActions gets the Application Auto Scaling scheduled actions for the given service namespace
func GetScheduledActions(t testing.TestingT, region string, serviceNamespace string) []types.ScheduledAction {
	actions, err := GetScheduledActionsE(t, region, serviceNamespace)
	require.NoError(t, err)
	return actions
}

// GetScheduledActionsE gets the Application Auto Scaling scheduled actions for the given service namespace
func GetScheduledActionsE(t testing.TestingT, region string, serviceNamespace string) ([]types.ScheduledAction, error) {
	client, err := NewApplicationAutoScalingClientE(t, region)
	if err != nil {
		return nil, err
	}

	result, err := client.DescribeScheduledActions(context.Background(), &applicationautoscaling.DescribeScheduledActionsInput{
		ServiceNamespace: types.ServiceNamespace(serviceNamespace),
	})

	if err != nil {
		return nil, err
	}

	return result.ScheduledActions, nil
}

// GetScalableTargetsByResourceId gets scalable targets filtered by resource ID
func GetScalableTargetsByResourceId(t testing.TestingT, region string, serviceNamespace string, resourceId string) []types.ScalableTarget {
	targets, err := GetScalableTargetsByResourceIdE(t, region, serviceNamespace, resourceId)
	require.NoError(t, err)
	return targets
}

// GetScalableTargetsByResourceIdE gets scalable targets filtered by resource ID
func GetScalableTargetsByResourceIdE(t testing.TestingT, region string, serviceNamespace string, resourceId string) ([]types.ScalableTarget, error) {
	client, err := NewApplicationAutoScalingClientE(t, region)
	if err != nil {
		return nil, err
	}

	result, err := client.DescribeScalableTargets(context.Background(), &applicationautoscaling.DescribeScalableTargetsInput{
		ServiceNamespace: types.ServiceNamespace(serviceNamespace),
		ResourceIds:      []string{resourceId},
	})

	if err != nil {
		return nil, err
	}

	return result.ScalableTargets, nil
}

// GetScheduledActionsByResourceId gets scheduled actions filtered by resource ID
func GetScheduledActionsByResourceId(t testing.TestingT, region string, serviceNamespace string, resourceId string) []types.ScheduledAction {
	actions, err := GetScheduledActionsByResourceIdE(t, region, serviceNamespace, resourceId)
	require.NoError(t, err)
	return actions
}

// GetScheduledActionsByResourceIdE gets scheduled actions filtered by resource ID
func GetScheduledActionsByResourceIdE(t testing.TestingT, region string, serviceNamespace string, resourceId string) ([]types.ScheduledAction, error) {
	client, err := NewApplicationAutoScalingClientE(t, region)
	if err != nil {
		return nil, err
	}

	result, err := client.DescribeScheduledActions(context.Background(), &applicationautoscaling.DescribeScheduledActionsInput{
		ServiceNamespace: types.ServiceNamespace(serviceNamespace),
		ResourceId:       aws.String(resourceId),
	})

	if err != nil {
		return nil, err
	}

	return result.ScheduledActions, nil
}

// NewApplicationAutoScalingClient creates a new Application Auto Scaling client
func NewApplicationAutoScalingClient(t testing.TestingT, region string) *applicationautoscaling.Client {
	client, err := NewApplicationAutoScalingClientE(t, region)
	require.NoError(t, err)
	return client
}

// NewApplicationAutoScalingClientE creates a new Application Auto Scaling client
func NewApplicationAutoScalingClientE(t testing.TestingT, region string) (*applicationautoscaling.Client, error) {
	sess, err := terratestaws.NewAuthenticatedSession(region)
	if err != nil {
		return nil, err
	}
	return applicationautoscaling.NewFromConfig(*sess), nil
}
