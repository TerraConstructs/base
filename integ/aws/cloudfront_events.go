package aws

import (
	"encoding/json"
	"fmt"
	"os"
)

type Version string

const Version1_0 Version = "1.0"

func (v *Version) UnmarshalJSON(data []byte) error {
	var versionStr string
	if err := json.Unmarshal(data, &versionStr); err != nil {
		return err
	}
	if versionStr != string(Version1_0) {
		return fmt.Errorf("unsupported version: %s", versionStr)
	}
	*v = Version(versionStr)
	return nil
}

func (v Version) MarshalJSON() ([]byte, error) {
	if v != Version1_0 {
		return nil, fmt.Errorf("unsupported version: %s", v)
	}
	return json.Marshal(string(v))
}

// CloudFrontFunctionEvent represents the event object that you provide to test a CloudFront function.
// https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/functions-event-structure.html
type CloudFrontFunctionEvent struct {
	Version  Version   `json:"version"` // The version field contains a string that specifies the version of the CloudFront Functions event object. The current version is 1.0.
	Context  Context   `json:"context"` // The context object contains contextual information about the event
	Viewer   Viewer    `json:"viewer"`
	Request  *Request  `json:"request,omitempty"`
	Response *Response `json:"response,omitempty"`
}

type Context struct {
	DistributionDomainName string `json:"distributionDomainName"` // The CloudFront domain name (for example, d111111abcdef8.cloudfront.net) of the distribution that's associated with the event.
	DistributionID         string `json:"distributionId"`
	EventType              string `json:"eventType"`
	RequestID              string `json:"requestId"`
}

type Viewer struct {
	IP string `json:"ip"`
}

type ValueObject map[string]ValueEntry

type ValueEntry struct {
	Value      string       `json:"value"`
	MultiValue []MultiValue `json:"multiValue,omitempty"`
}

type MultiValue struct {
	Value string `json:"value"`
}

type Request struct {
	Method      string      `json:"method"`
	URI         string      `json:"uri"`
	Querystring ValueObject `json:"querystring"`
	Headers     ValueObject `json:"headers"`
	Cookies     ValueObject `json:"cookies,omitempty"`
}

type Response struct {
	StatusCode        int             `json:"statusCode"`
	StatusDescription *string         `json:"statusDescription,omitempty"`
	Headers           *ValueObject    `json:"headers,omitempty"`
	Cookies           *ResponseCookie `json:"cookies,omitempty"`
}

type ResponseCookie map[string]ResponseCookieEntry

type ResponseCookieEntry struct {
	Value      string            `json:"value"`
	Attributes string            `json:"attributes"`
	MultiValue []MultiValueAttrs `json:"multiValue,omitempty"`
}

type MultiValueAttrs struct {
	Value      string `json:"value"`
	Attributes string `json:"attributes"`
}

func ReadCloudFrontEvent(path string) (*CloudFrontFunctionEvent, error) {
	f, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}

	var t = CloudFrontFunctionEvent{
		Version: "1.0",
		Context: Context{
			EventType: "viewer-request",
		},
		Viewer: Viewer{
			IP: "1.2.3.4",
		},
	}

	err = json.Unmarshal(f, &t)
	return &t, err
}
