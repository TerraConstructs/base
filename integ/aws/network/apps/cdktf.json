{
  "//": "Dummy file to satisfy the bundling findUp logic"
}
