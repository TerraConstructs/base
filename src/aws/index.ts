// core
// export * from "./tag-aspect";
// export * from "./tag-manager";
export * from "./arn";
export * from "./aws-construct";
export * from "./aws-stack";
export * from "./aws-tags";
export * from "./log-retention";

// core JSII compatible generated interfaces
export * from "./provider-config.generated";

// classes
export * as iam from "./iam";
export * as encryption from "./encryption";
export * as compute from "./compute";
export * as network from "./network";
export * as storage from "./storage";
export * as notify from "./notify";
export * as edge from "./edge";
export * as cloudwatch from "./cloudwatch";
