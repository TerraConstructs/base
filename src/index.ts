export * from "./construct-base";
export * from "./stack-base";
export * from "./duration";
export * from "./size";

// cloud specific exports
export * as aws from "./aws";
