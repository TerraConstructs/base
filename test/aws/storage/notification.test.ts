import { App, Testing } from "cdktf";
import "cdktf/lib/testing/adapters/jest";
import { storage, AwsSpec, iam } from "../../../src/aws";

const environmentName = "Test";
const gridUUID = "123e4567-e89b-12d3";
const providerConfig = { region: "us-east-1" };
const gridBackendConfig = {
  address: "http://localhost:3000",
};

describe("notification", () => {
  let app: App;
  let spec: AwsSpec;

  beforeEach(() => {
    app = Testing.app();
    spec = new AwsSpec(app, "MyStack", {
      environmentName,
      gridUUID,
      providerConfig,
      gridBackendConfig,
      // TODO: Should support passing account via Spec props?
      // account: "1234",
      // region: "us-east-1",
    });
  });

  test("when notification is added a custom s3 bucket notification resource is provisioned", () => {
    // GIVEN
    const bucket = new storage.Bucket(spec, "MyBucket");

    // WHEN
    bucket.addEventNotification(storage.EventType.OBJECT_CREATED, {
      bind: () => ({
        arn: "ARN",
        type: storage.BucketNotificationDestinationType.TOPIC,
      }),
    });

    // THEN

    // Do prepare run to resolve/add all Terraform resources
    spec.prepareStack();
    const synthesized = Testing.synth(spec);
    // refer to full snapshot for debug
    expect(synthesized).toMatchSnapshot();
    // Template.fromStack(spec).resourceCountIs("AWS::S3::Bucket", 1);
    // Template.fromStack(spec).hasResourceProperties(
    //   "Custom::S3BucketNotifications",
    //   {
    //     NotificationConfiguration: {
    //       TopicConfigurations: [
    //         {
    //           Events: ["s3:ObjectCreated:*"],
    //           TopicArn: "ARN",
    //         },
    //       ],
    //     },
    //   },
    // );
  });

  // test("can specify a custom role for the notifications handler of imported buckets", () => {
  //   const importedRole = iam.Role.fromRoleArn(
  //     spec,
  //     "role",
  //     "arn:aws:iam::111111111111:role/DevsNotAllowedToTouch",
  //   );

  //   const bucket = storage.Bucket.fromBucketAttributes(spec, "MyBucket", {
  //     bucketName: "foo-bar",
  //     notificationsHandlerRole: importedRole,
  //   });

  //   bucket.addEventNotification(storage.EventType.OBJECT_CREATED, {
  //     bind: () => ({
  //       arn: "ARN",
  //       type: storage.BucketNotificationDestinationType.TOPIC,
  //     }),
  //   });

  //   // Do prepare run to resolve/add all Terraform resources
  //   spec.prepareStack();
  //   const synthesized = Testing.synth(spec);
  //   // refer to full snapshot for debug
  //   expect(synthesized).toMatchSnapshot();
  //   // Template.fromStack(spec).hasResourceProperties("AWS::Lambda::Function", {
  //   //   Description:
  //   //     'AWS CloudFormation handler for "Custom::S3BucketNotifications" resources (@aws-cdk/aws-s3)',
  //   //   Role: "arn:aws:iam::111111111111:role/DevsNotAllowedToTouch",
  //   // });
  // });

  test("can specify prefix and suffix filter rules", () => {
    // GIVEN
    const bucket = new storage.Bucket(spec, "MyBucket");

    // WHEN
    bucket.addEventNotification(
      storage.EventType.OBJECT_CREATED,
      {
        bind: () => ({
          arn: "ARN",
          type: storage.BucketNotificationDestinationType.TOPIC,
        }),
      },
      { prefix: "images/", suffix: ".png" },
    );

    // THEN
    // Do prepare run to resolve/add all Terraform resources
    spec.prepareStack();
    const synthesized = Testing.synth(spec);
    // refer to full snapshot for debug
    expect(synthesized).toMatchSnapshot();
    // Template.fromStack(spec).hasResourceProperties(
    //   "Custom::S3BucketNotifications",
    //   {
    //     NotificationConfiguration: {
    //       TopicConfigurations: [
    //         {
    //           Events: ["s3:ObjectCreated:*"],
    //           Filter: {
    //             Key: {
    //               FilterRules: [
    //                 {
    //                   Name: "suffix",
    //                   Value: ".png",
    //                 },
    //                 {
    //                   Name: "prefix",
    //                   Value: "images/",
    //                 },
    //               ],
    //             },
    //           },
    //           TopicArn: "ARN",
    //         },
    //       ],
    //     },
    //   },
    // );
  });

  // TODO: We do not use custom handler to manage bucket notifications
  test.skip("the notification lambda handler must depend on the role to prevent executing too early", () => {
    // GIVEN
    const bucket = new storage.Bucket(spec, "MyBucket");

    // WHEN
    bucket.addEventNotification(storage.EventType.OBJECT_CREATED, {
      bind: () => ({
        arn: "ARN",
        type: storage.BucketNotificationDestinationType.TOPIC,
      }),
    });

    // THEN
    // Do prepare run to resolve/add all Terraform resources
    spec.prepareStack();
    const synthesized = Testing.synth(spec);
    // refer to full snapshot for debug
    expect(synthesized).toMatchSnapshot();
    // Template.fromStack(spec).hasResource("AWS::Lambda::Function", {
    //   Type: "AWS::Lambda::Function",
    //   Properties: {
    //     Role: {
    //       "Fn::GetAtt": [
    //         "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleB6FB88EC",
    //         "Arn",
    //       ],
    //     },
    //   },
    //   DependsOn: [
    //     "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleDefaultPolicy2CF63D36",
    //     "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleB6FB88EC",
    //   ],
    // });
  });

  test("must not depend on bucket policy if bucket policy does not exists", () => {
    const bucket = new storage.Bucket(spec, "MyBucket");

    bucket.addEventNotification(storage.EventType.OBJECT_CREATED, {
      bind: () => ({
        arn: "ARN",
        type: storage.BucketNotificationDestinationType.TOPIC,
      }),
    });

    // Do prepare run to resolve/add all Terraform resources
    spec.prepareStack();
    const synthesized = Testing.synth(spec);
    // refer to full snapshot for debug
    // expect(synthesized).toMatchSnapshot();
    expect(
      JSON.parse(synthesized).resource.aws_s3_bucket_notification
        .MyBucket_Notifications_46AC0CD2,
    ).not.toHaveProperty("depends_on");
    // Template.fromStack(spec).hasResource("Custom::S3BucketNotifications", {
    //   Type: "Custom::S3BucketNotifications",
    //   DependsOn: Match.absent(),
    // });
  });

  test("must depend on bucket policy to prevent executing too early", () => {
    const bucket = new storage.Bucket(spec, "MyBucket", {
      enforceSSL: true, // adds bucket policy for test
    });

    bucket.addEventNotification(storage.EventType.OBJECT_CREATED, {
      bind: () => ({
        arn: "ARN",
        type: storage.BucketNotificationDestinationType.TOPIC,
      }),
    });

    // Do prepare run to resolve/add all Terraform resources
    spec.prepareStack();
    const synthesized = Testing.synth(spec);
    // refer to full snapshot for debug
    // expect(synthesized).toMatchSnapshot();
    const template = JSON.parse(synthesized);
    expect(template).toMatchObject({
      resource: {
        aws_s3_bucket_notification: {
          MyBucket_Notifications_46AC0CD2: {
            depends_on: [
              "data.aws_iam_policy_document.MyBucket_Policy_F89E7330",
              "aws_s3_bucket_policy.MyBucket_Policy_E7FBAC7B",
            ],
          },
        },
      },
    });
    // Template.fromStack(spec).hasResource("Custom::S3BucketNotifications", {
    //   Type: "Custom::S3BucketNotifications",
    //   DependsOn: ["MyBucketPolicyE7FBAC7B"],
    // });
  });

  test("must depend on bucket policy even if bucket policy is added after notification", () => {
    const bucket = new storage.Bucket(spec, "MyBucket");

    bucket.addEventNotification(storage.EventType.OBJECT_CREATED, {
      bind: () => ({
        arn: "ARN",
        type: storage.BucketNotificationDestinationType.TOPIC,
      }),
    });

    bucket.addToResourcePolicy(
      new iam.PolicyStatement({
        resources: [bucket.bucketArn],
        actions: ["s3:GetBucketAcl"],
        principals: [new iam.AnyPrincipal()],
      }),
    );

    // Do prepare run to resolve/add all Terraform resources
    spec.prepareStack();
    const synthesized = Testing.synth(spec);
    // refer to full snapshot for debug
    // expect(synthesized).toMatchSnapshot();
    const template = JSON.parse(synthesized);
    expect(template).toMatchObject({
      resource: {
        aws_s3_bucket_notification: {
          MyBucket_Notifications_46AC0CD2: {
            depends_on: [
              "data.aws_iam_policy_document.MyBucket_Policy_F89E7330",
              "aws_s3_bucket_policy.MyBucket_Policy_E7FBAC7B",
            ],
          },
        },
      },
    });
  });

  // TODO: Terraform doesn't have this limitation?
  test("does not throw if both prefix or suffix set for a filter", () => {
    const bucket = new storage.Bucket(spec, "MyBucket");

    bucket.addEventNotification(
      storage.EventType.OBJECT_CREATED,
      {
        bind: () => ({
          arn: "ARN",
          type: storage.BucketNotificationDestinationType.TOPIC,
        }),
      },
      { prefix: "foo/", suffix: "bar/" },
    );
    // Do prepare run to resolve/add all Terraform resources
    spec.prepareStack();
    const synthesized = Testing.synth(spec);
    // refer to full snapshot for debug
    // expect(synthesized).toMatchSnapshot();
    const template = JSON.parse(synthesized);
    expect(template).toMatchObject({
      resource: {
        aws_s3_bucket_notification: {
          MyBucket_Notifications_46AC0CD2: {
            bucket: "${aws_s3_bucket.MyBucket_F68F3FF0.bucket}",
            eventbridge: false,
            topic: [
              {
                events: ["s3:ObjectCreated:*"],
                filter_prefix: "foo/",
                filter_suffix: "bar/",
                topic_arn: "ARN",
              },
            ],
          },
        },
      },
    });
    // expect(() =>
    //   bucket.addEventNotification(
    //     storage.EventType.OBJECT_CREATED,
    //     {
    //       bind: () => ({
    //         arn: "ARN",
    //         type: storage.BucketNotificationDestinationType.TOPIC,
    //       }),
    //     },
    //     { prefix: "foo/", suffix: "bar/" },
    //   ),
    // ).toThrow(/`prefix` and\/or `suffix`/);
  });

  // TODO: Terraform doesn't have this limitation?
  test("does not throw  with multiple prefix rules in a filter", () => {
    const bucket = new storage.Bucket(spec, "MyBucket");
    bucket.addEventNotification(
      storage.EventType.OBJECT_CREATED,
      {
        bind: () => ({
          arn: "ARN",
          type: storage.BucketNotificationDestinationType.TOPIC,
        }),
      },
      { prefix: "images/" },
      { prefix: "archive/" },
    );
    // Do prepare run to resolve/add all Terraform resources
    spec.prepareStack();
    const synthesized = Testing.synth(spec);
    // refer to full snapshot for debug
    // expect(synthesized).toMatchSnapshot();
    const template = JSON.parse(synthesized);
    expect(template).toMatchObject({
      resource: {
        aws_s3_bucket_notification: {
          MyBucket_Notifications_46AC0CD2: {
            bucket: "${aws_s3_bucket.MyBucket_F68F3FF0.bucket}",
            eventbridge: false,
            topic: [
              {
                events: ["s3:ObjectCreated:*"],
                filter_prefix: "images/",
                topic_arn: "ARN",
              },
              {
                events: ["s3:ObjectCreated:*"],
                filter_prefix: "archive/",
                topic_arn: "ARN",
              },
            ],
          },
        },
      },
    });
    // expect(() =>
    //   bucket.addEventNotification(
    //     storage.EventType.OBJECT_CREATED,
    //     {
    //       bind: () => ({
    //         arn: "ARN",
    //         type: storage.BucketNotificationDestinationType.TOPIC,
    //       }),
    //     },
    //     { prefix: "images/" },
    //     { prefix: "archive/" },
    //   ),
    // ).toThrow(/prefix rule/);
  });

  // TODO: Terraform doesn't have this limitation?
  test("does not throw with multiple suffix rules in a filter", () => {
    const bucket = new storage.Bucket(spec, "MyBucket");

    bucket.addEventNotification(
      storage.EventType.OBJECT_CREATED,
      {
        bind: () => ({
          arn: "ARN",
          type: storage.BucketNotificationDestinationType.TOPIC,
        }),
      },
      { suffix: ".png" },
      { suffix: ".zip" },
    );
    // Do prepare run to resolve/add all Terraform resources
    spec.prepareStack();
    const synthesized = Testing.synth(spec);
    // refer to full snapshot for debug
    // expect(synthesized).toMatchSnapshot();
    const template = JSON.parse(synthesized);
    expect(template).toMatchObject({
      resource: {
        aws_s3_bucket_notification: {
          MyBucket_Notifications_46AC0CD2: {
            bucket: "${aws_s3_bucket.MyBucket_F68F3FF0.bucket}",
            eventbridge: false,
            topic: [
              {
                events: ["s3:ObjectCreated:*"],
                filter_suffix: ".png",
                topic_arn: "ARN",
              },
              {
                events: ["s3:ObjectCreated:*"],
                filter_suffix: ".zip",
                topic_arn: "ARN",
              },
            ],
          },
        },
      },
    });
    // expect(() =>
    //   bucket.addEventNotification(
    //     storage.EventType.OBJECT_CREATED,
    //     {
    //       bind: () => ({
    //         arn: "ARN",
    //         type: storage.BucketNotificationDestinationType.TOPIC,
    //       }),
    //     },
    //     { suffix: ".png" },
    //     { suffix: ".zip" },
    //   ),
    // ).toThrow(/suffix rule/);
  });

  test("EventBridge notification resource", () => {
    // WHEN
    new storage.Bucket(spec, "MyBucket", {
      eventBridgeEnabled: true,
    });

    // THEN
    // Do prepare run to resolve/add all Terraform resources
    spec.prepareStack();
    const synthesized = Testing.synth(spec);
    // refer to full snapshot for debug
    // expect(synthesized).toMatchSnapshot();
    const template = JSON.parse(synthesized);
    expect(template).toMatchObject({
      resource: {
        aws_s3_bucket_notification: {
          MyBucket_Notifications_46AC0CD2: {
            bucket: "${aws_s3_bucket.MyBucket_F68F3FF0.bucket}",
            eventbridge: true,
          },
        },
      },
    });
    // Template.fromStack(spec).resourceCountIs("AWS::S3::Bucket", 1);
    // Template.fromStack(spec).hasResourceProperties(
    //   "Custom::S3BucketNotifications",
    //   {
    //     NotificationConfiguration: {
    //       EventBridgeConfiguration: {},
    //     },
    //   },
    // );
  });

  // test("skip destination validation is set to false by default", () => {
  //   // GIVEN
  //   const stack = new cdk.Stack();

  //   // WHEN
  //   const bucket = new storage.Bucket(stack, "MyBucket", {
  //     bucketName: "foo-bar",
  //   });
  //   bucket.addEventNotification(storage.EventType.OBJECT_CREATED, {
  //     bind: () => ({
  //       arn: "ARN",
  //       type: storage.BucketNotificationDestinationType.TOPIC,
  //     }),
  //   });

  //   // THEN
  //   Template.fromStack(stack).hasResourceProperties(
  //     "Custom::S3BucketNotifications",
  //     {
  //       SkipDestinationValidation: false,
  //     },
  //   );
  // });
});
