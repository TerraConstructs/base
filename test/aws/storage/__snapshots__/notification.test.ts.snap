// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`notification can specify prefix and suffix filter rules 1`] = `
"{
  "data": {
    "aws_caller_identity": {
      "CallerIdentity": {
        "provider": "aws"
      }
    },
    "aws_partition": {
      "Partitition": {
        "provider": "aws"
      }
    }
  },
  "provider": {
    "aws": [
      {
        "region": "us-east-1"
      }
    ]
  },
  "resource": {
    "aws_s3_bucket": {
      "MyBucket_F68F3FF0": {
        "bucket_prefix": "123e4567-e89b-12d3-mystackmybucket",
        "tags": {
          "Name": "Test-MyBucket",
          "et:grid:EnvironmentName": "Test",
          "et:grid:UUID": "123e4567-e89b-12d3"
        }
      }
    },
    "aws_s3_bucket_notification": {
      "MyBucket_Notifications_46AC0CD2": {
        "bucket": "\${aws_s3_bucket.MyBucket_F68F3FF0.bucket}",
        "eventbridge": false,
        "topic": [
          {
            "events": [
              "s3:ObjectCreated:*"
            ],
            "filter_prefix": "images/",
            "filter_suffix": ".png",
            "topic_arn": "ARN"
          }
        ]
      }
    }
  },
  "terraform": {
    "backend": {
      "http": {
        "address": "http://localhost:3000"
      }
    },
    "required_providers": {
      "aws": {
        "source": "aws",
        "version": "5.68.0"
      }
    }
  }
}"
`;

exports[`notification when notification is added a custom s3 bucket notification resource is provisioned 1`] = `
"{
  "data": {
    "aws_caller_identity": {
      "CallerIdentity": {
        "provider": "aws"
      }
    },
    "aws_partition": {
      "Partitition": {
        "provider": "aws"
      }
    }
  },
  "provider": {
    "aws": [
      {
        "region": "us-east-1"
      }
    ]
  },
  "resource": {
    "aws_s3_bucket": {
      "MyBucket_F68F3FF0": {
        "bucket_prefix": "123e4567-e89b-12d3-mystackmybucket",
        "tags": {
          "Name": "Test-MyBucket",
          "et:grid:EnvironmentName": "Test",
          "et:grid:UUID": "123e4567-e89b-12d3"
        }
      }
    },
    "aws_s3_bucket_notification": {
      "MyBucket_Notifications_46AC0CD2": {
        "bucket": "\${aws_s3_bucket.MyBucket_F68F3FF0.bucket}",
        "eventbridge": false
      }
    }
  },
  "terraform": {
    "backend": {
      "http": {
        "address": "http://localhost:3000"
      }
    },
    "required_providers": {
      "aws": {
        "source": "aws",
        "version": "5.68.0"
      }
    }
  }
}"
`;
